weave/integrations/google_genai/gemini_utils.py

```diff
@@ -0,0 +1,148 @@
+        return value
+
+    for i, value_candidate in enumerate(value.candidates):
+        if i >= len(acc.candidates):
+            break
+        for j, value_part in enumerate(value_candidate.content.parts):
+            if j >= len(acc.candidates[i].content.parts):
+                break
+            if value_part.text is not None:
+                acc.candidates[i].content.parts[j].text += value_part.text
+
+    if acc.usage_metadata.prompt_token_count is None:
+        acc.usage_metadata.prompt_token_count = 0
+    elif value.usage_metadata.prompt_token_count is not None:
+        acc.usage_metadata.prompt_token_count += value.usage_metadata.prompt_token_count
+
+    if acc.usage_metadata.candidates_token_count is None:
+        acc.usage_metadata.candidates_token_count = 0
+    elif value.usage_metadata.candidates_token_count is not None:
+        acc.usage_metadata.candidates_token_count += (
+            value.usage_metadata.candidates_token_count
+        )
+
+    if acc.usage_metadata.total_token_count is None:
+        acc.usage_metadata.total_token_count = 0
+    elif value.usage_metadata.total_token_count is not None:
+        acc.usage_metadata.total_token_count += value.usage_metadata.total_token_count
+
+    if acc.usage_metadata.cached_content_token_count is None:
+        acc.usage_metadata.cached_content_token_count = 0
+    elif value.usage_metadata.cached_content_token_count is not None:
+        acc.usage_metadata.cached_content_token_count += (
+            value.usage_metadata.cached_content_token_count
+        )
+
+    return acc
+
+
+def google_genai_gemini_wrapper_sync(
+    settings: OpSettings,
+) -> Callable[[Callable], Callable]:
+    def wrapper(fn: Callable) -> Callable:
+        op_kwargs = settings.model_dump()
+        if not op_kwargs.get("postprocess_inputs"):
+            op_kwargs["postprocess_inputs"] = google_genai_gemini_postprocess_inputs
+
+        op = weave.op(fn, **op_kwargs)
+        if not op.name.endswith("count_tokens"):
+            op._set_on_finish_handler(google_genai_gemini_on_finish)
+        return add_accumulator(
+            op,
+            make_accumulator=lambda inputs: google_genai_gemini_accumulator,
+            should_accumulate=lambda inputs: op.name.endswith("stream"),
+        )
+
+    return wrapper
+
+
+def google_genai_gemini_wrapper_async(
+    settings: OpSettings,
+) -> Callable[[Callable], Callable]:
+    def wrapper(fn: Callable) -> Callable:
+        def _fn_wrapper(fn: Callable) -> Callable:
+            @wraps(fn)
+            async def _async_wrapper(*args: Any, **kwargs: Any) -> Any:
+                return await fn(*args, **kwargs)
+
+            return _async_wrapper
+
+        op_kwargs = settings.model_dump()
+        if not op_kwargs.get("postprocess_inputs"):
+            op_kwargs["postprocess_inputs"] = google_genai_gemini_postprocess_inputs
+
+        op = weave.op(_fn_wrapper(fn), **op_kwargs)
+        if not op.name.endswith("count_tokens"):
```

Comment:
is there some special op with this?

Reply:
Yes check https://github.com/wandb/weave/pull/3596/#discussion_r2004799294

Reply:
See my comment above.  It will make this path more consistent in case we need to update the list of fns in future